cosmos/dbt/parser/project.py

```diff
@@ -87,6 +94,30 @@ def _config_selector_ooo(
         return sql_configs
 
 
+def extract_python_file_upstream_requirements(code: str) -> list[str]:
+    """
+    Given a dbt Python model source code, identify other dbt entities which are upstream requirements.
+
+    This method tries to find a `model` function and `dbt.ref` calls within it. Return a list of upstream entities IDs.
+    """
+    source_code = ast.parse(code)
+
+    upstream_entities = []
+    model_function = ""
+    for node in source_code.body:
+        if isinstance(node, ast.FunctionDef) and node.name == DBT_PY_MODEL_METHOD_NAME:
+            model_function = node
+            break
+
+    for item in ast.walk(model_function):
+        if isinstance(item, ast.Call) and item.func.attr == DBT_PY_DEP_METHOD_NAME:
```

Comment:
how about sources? https://docs.getdbt.com/docs/build/python-models#referencing-other-models

Reply:
Ah, I missed this! Will add support now.

I think there may be a few circumstances in which we may miss references to upstream Python models, e.g. if they are in it is in an f-string. But we can iterate over this feature over time based on user's feedback!

Reply:
Ahh right we've had conversations about this before, thanks for reminding me. This is good as-is.